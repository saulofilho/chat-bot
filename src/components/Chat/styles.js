import styled from 'styled-components';

export const ChatToggle = styled.div`
  cursor: pointer;
  width: 60px;
  height: 60px;
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 50%;
  border: 2px solid black;
  background: rgb(34, 193, 195);
  background: linear-gradient(
    0deg,
    rgba(34, 193, 195, 1) 0%,
    rgba(253, 187, 45, 1) 100%
  );
  margin: 0 auto 10px;
`;

export const Content = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  flex-flow: column;
  display: ${props => (props.hide ? 'flex' : 'none')};

  form {
    display: flex;
    overflow: scroll;
    height: 500px;
    justify-content: flex-start;
    align-items: center;
    flex-flow: column;
    background: rgb(34, 193, 195);
    background: linear-gradient(
      0deg,
      rgba(34, 193, 195, 1) 0%,
      rgba(253, 187, 45, 1) 100%
    );
    border: 2px solid #000;
    border-radius: 2%;
    padding: 20px;
    margin: 0 auto 20px;
  }
`;

export const SpeechBubble = styled.div`
  position: relative;
  background: #fff;
  border-radius: 0.4em;
  width: 200px;
  height: auto;

  :after {
    content: '';
    position: absolute;
    left: 0;
    top: 50%;
    width: 0;
    height: 0;
    border: 22px solid transparent;
    border-right-color: #fff;
    border-left: 0;
    border-top: 0;
    margin-top: -11px;
    margin-left: -22px;
  }
`;
