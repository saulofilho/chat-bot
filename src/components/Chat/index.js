import React, { useState } from 'react';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import { messages } from '../../utils/messages';

import { ChatToggle, Content, SpeechBubble } from './styles';

function Chat() {
  const [isToggled, setIsToggled] = useState(false);

  const toggle = () => {
    setIsToggled(!isToggled);
  };

  const formik = useFormik({
    initialValues: {
      fullName: '',
      city: '',
      birth: '',
      email: '',
    },
    validationSchema: Yup.object({
      fullName: Yup.string()
        .max(15, 'Must be 15 characters or less')
        .required('Required'),
      city: Yup.string()
        .min(3, 'Must be a valid city')
        .required('Required'),
      birth: Yup.number()
        .lessThan(8, 'Invalid birth number')
        .required('Required'),
      email: Yup.string()
        .email('Invalid email address')
        .required('Required'),
      rating: Yup.number()
        .lessThan(5, 'Invalid rating')
        .required('Required'),
    }),
    onSubmit: values => {
      alert(JSON.stringify(values, null, 2));
    },
  });

  return (
    <>
      <ChatToggle onClick={() => toggle()}>
        {!isToggled ? 'Olá' : 'x'}
      </ChatToggle>
      <Content hide={isToggled}>
        <form onSubmit={formik.handleSubmit}>
          {messages.map(item => {
            return (
              <div key={item.id}>
                {item.id === 1 ? (
                  <div>
                    <p>🤖</p>
                    <SpeechBubble>{item.message}</SpeechBubble>
                  </div>
                ) : null}
              </div>
            );
          })}
          <label htmlFor="fullName">
            First Name
            <input
              id="fullName"
              name="fullName"
              type="text"
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              value={formik.values.fullName}
            />
          </label>
          {formik.touched.fullName && formik.errors.fullName ? (
            <div>{formik.errors.fullName}</div>
          ) : null}

          {messages.map(item => {
            return (
              <div key={item.id}>
                {item.id === 2 ? (
                  <div>
                    <p>🤖</p>
                    <SpeechBubble>{item.message}</SpeechBubble>
                  </div>
                ) : null}
              </div>
            );
          })}
          <label htmlFor="city">
            City
            <input
              id="city"
              name="city"
              type="text"
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              value={formik.values.city}
            />
          </label>
          {formik.touched.city && formik.errors.city ? (
            <div>{formik.errors.city}</div>
          ) : null}

          {messages.map(item => {
            return (
              <div key={item.id}>
                {item.id === 3 ? (
                  <div>
                    <p>🤖</p>
                    <SpeechBubble>{item.message}</SpeechBubble>
                  </div>
                ) : null}
              </div>
            );
          })}
          <label htmlFor="birth">
            Birth
            <input
              id="birth"
              name="birth"
              type="birth"
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              value={formik.values.birth}
            />
          </label>
          {formik.touched.birth && formik.errors.birth ? (
            <div>{formik.errors.birth}</div>
          ) : null}

          {messages.map(item => {
            return (
              <div key={item.id}>
                {item.id === 4 ? (
                  <div>
                    <p>🤖</p>
                    <SpeechBubble>{item.message}</SpeechBubble>
                  </div>
                ) : null}
              </div>
            );
          })}
          <label htmlFor="email">
            Email Address
            <input
              id="email"
              name="email"
              type="email"
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              value={formik.values.email}
            />
          </label>
          {formik.touched.email && formik.errors.email ? (
            <div>{formik.errors.email}</div>
          ) : null}

          {messages.map(item => {
            return (
              <div key={item.id}>
                {item.id === 5 ? (
                  <div>
                    <p>🤖</p>
                    <SpeechBubble>{item.message}</SpeechBubble>
                  </div>
                ) : null}
              </div>
            );
          })}
          <label htmlFor="rating">
            Rating
            <input
              id="rating"
              name="rating"
              type="rating"
              onChange={formik.handleChange}
              onBlur={formik.handleBlur}
              value={formik.values.rating}
            />
          </label>
          {formik.touched.rating && formik.errors.rating ? (
            <div>{formik.errors.rating}</div>
          ) : null}

          <button type="submit">Submit</button>
        </form>
      </Content>
    </>
  );
}

export default Chat;
