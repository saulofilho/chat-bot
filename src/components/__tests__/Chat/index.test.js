import React from 'react';
import { render, cleanup } from '@testing-library/react';
import axios from 'axios';

import Chat from '../../Chat';

jest.mock('axios');

describe('testing chat component', () => {
  it('should render correctly', () => {
    const wrapper = render(<Chat />);

    expect(wrapper).toMatchSnapshot();
  });
});

afterEach(cleanup);

test('testing toggle function', () => {
  const { getByTestId } = render(<Chat />);
  expect(getByTestId('div-toggle')).toHaveTextContent('Olá' || 'Tchau');
});

test('should fetch city', async () => {
  async function apiBrazilCity() {
    const response = await axios.get(
      'https://servicodados.ibge.gov.br/api/v1/localidades/municipios'
    );

    return response.data[0].nome;
  }

  await axios.get.mockResolvedValue({
    data: [
      {
        id: 1100049,
        nome: 'Cacoal',
        microrregiao: {
          id: 11006,
          nome: 'Cacoal',
          mesorregiao: {
            id: 1102,
            nome: 'Leste Rondoniense',
            UF: {
              id: 11,
              sigla: 'RO',
              nome: 'Rondônia',
              regiao: {
                id: 1,
                sigla: 'N',
                nome: 'Norte',
              },
            },
          },
        },
      },
    ],
  });

  const nome = await apiBrazilCity();
  await expect(nome).toEqual('Cacoal');
});

test('should post user', async () => {
  async function onSubmit(values) {
    await axios.post(
      'https://6058bdc9c3f49200173aecbc.mockapi.io/users',
      values
    );
  }

  await axios.post.mockResolvedValue({
    values: {
      fullName: 'Teste Jr.',
      city: 'Testeland',
      birth: '01/01/2021',
      email: 'teste@teste.com',
      rating: '5',
    },
  });

  const data = await onSubmit();
  await expect(data).toEqual(data);
});
