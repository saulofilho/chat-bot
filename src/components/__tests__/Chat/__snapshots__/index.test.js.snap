// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing chat component should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdfBwQ gzzPqc"
        data-testid="div-toggle"
      >
        Ol√°
      </div>
      <div
        class="sc-gsTCUz koRIYN"
      >
        <form>
          <div>
            <div
              class="sc-hKgILt hWEPCc"
            >
              <div
                class="sc-eCssSg eKPVDe"
              >
                ü§ñ
              </div>
              <div
                class="sc-dlfnbm eSMgzp"
              >
                Ol√°, eu sou Chatnilson, tudo bem? Para come√ßarmos, preciso saber seu nome.
              </div>
            </div>
          </div>
          <div />
          <div />
          <div />
          <div />
          <div
            class="sc-gKsewC gxKhoi"
          >
            <div
              class="sc-fubCfw hQvcvs"
            >
              <input
                id="fullName"
                name="fullName"
                placeholder="Nome e sobrenome"
                type="text"
                value=""
              />
            </div>
            <div
              class="sc-iBPRYJ kEZjJC"
            />
          </div>
          <button
            class="sc-jSgupP chrnNF"
            type="submit"
          >
            Salvar
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdfBwQ gzzPqc"
      data-testid="div-toggle"
    >
      Ol√°
    </div>
    <div
      class="sc-gsTCUz koRIYN"
    >
      <form>
        <div>
          <div
            class="sc-hKgILt hWEPCc"
          >
            <div
              class="sc-eCssSg eKPVDe"
            >
              ü§ñ
            </div>
            <div
              class="sc-dlfnbm eSMgzp"
            >
              Ol√°, eu sou Chatnilson, tudo bem? Para come√ßarmos, preciso saber seu nome.
            </div>
          </div>
        </div>
        <div />
        <div />
        <div />
        <div />
        <div
          class="sc-gKsewC gxKhoi"
        >
          <div
            class="sc-fubCfw hQvcvs"
          >
            <input
              id="fullName"
              name="fullName"
              placeholder="Nome e sobrenome"
              type="text"
              value=""
            />
          </div>
          <div
            class="sc-iBPRYJ kEZjJC"
          />
        </div>
        <button
          class="sc-jSgupP chrnNF"
          type="submit"
        >
          Salvar
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
